lib/srv/app/server.go

```diff
@@ -542,15 +552,33 @@ func (s *Server) ForceHeartbeat() error {
 }
 
 // HandleConnection takes a connection and wraps it in a listener so it can
-// be passed to http.Serve to process as a HTTP request.
+// be passed to http.Serve to process as a HTTP request. This method blocks
+// until the connection closes.
 func (s *Server) HandleConnection(conn net.Conn) {
-	// Wrap the listener in a TLS authorizing listener.
-	listener := newListener(s.closeContext, conn)
-	tlsListener := tls.NewListener(listener, s.tlsConfig)
-
-	if err := s.httpServer.Serve(tlsListener); err != nil {
+	// Wrap the TLS authorizing conn in a single-use listener.
+	tlsConn := tls.Server(conn, s.tlsConfig)
+	listener := newListener(s.closeContext, tlsConn)
+
+	// build a channel that will close when tlsConn is closed
+	closeCh := s.httpConnState.Channel(tlsConn, http.StateClosed)
+
+	// Serve will return as soon as tlsConn is running in its own goroutine
+	err := s.httpServer.Serve(listener)
+	if err != nil && !errors.Is(err, errListenerConnServed) {
+		// okay to ignore errListenerConnServed; it is a signal that our
+		// single-use listener has passed the connection to http.Serve
+		// and conn is being served. See listener.Accept for details.
 		s.log.Warnf("Failed to handle connection: %v.", err)
+		s.httpConnState.Release(tlsConn)
+		return
+	}
+
+	// block until conn or Server is closed
```

Comment:
Nit: I think the comments should explain "why" rather than (or in addition to) "what". I.e. here I would explain why we're blocking here and what would happen if we don't.

Reply:
Updated